pkg/engine/jmespath/arithmetic.go

```diff
@@ -0,0 +1,423 @@
+package jmespath
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"time"
+
+	"gopkg.in/inf.v0"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type Operand interface {
+	Add(interface{}) (interface{}, error)
+	Subtract(interface{}) (interface{}, error)
+	Multiply(interface{}) (interface{}, error)
+	Divide(interface{}) (interface{}, error)
+	Modulo(interface{}) (interface{}, error)
+}
+
+type Quantity struct {
+	resource.Quantity
+}
+
+type Duration struct {
+	time.Duration
+}
+
+type Scalar struct {
+	float64
+}
+
+func ParseArithemticOperands(arguments []interface{}, operator string) (Operand, Operand, error) {
+	op := [2]Operand{nil, nil}
+	t := [2]int{0, 0}
+
+	for i := 0; i < 2; i++ {
+		tmp, err := validateArg(divide, arguments, i, reflect.Float64)
+		if err == nil {
+			var sc Scalar
+			sc.float64 = tmp.Float()
+			op[i] = sc
+		}
+
+		tmp, err = validateArg(divide, arguments, i, reflect.String)
+		if err == nil {
+			var q Quantity
+			q.Quantity, err = resource.ParseQuantity(tmp.String())
+			if err == nil {
+				op[i] = q
+				t[i] = 1
+			} else {
+				var d Duration
+				d.Duration, err = time.ParseDuration(tmp.String())
+				if err == nil {
+					op[i] = d
+					t[i] = 2
+				}
+			}
+		}
+	}
+
+	if op[0] == nil || op[1] == nil || t[0]|t[1] == 3 {
+		return nil, nil, fmt.Errorf(genericError, operator, "invalid operands")
+	}
+
+	return op[0], op[1], nil
+}
+
+func (op1 Quantity) Add(op2 interface{}) (interface{}, error) {
+	switch v := op2.(type) {
+	case Quantity:
+		op1.Quantity.Add(v.Quantity)
+		return op1.String(), nil
+	case Scalar:
```

Comment:
I generally dislike that we give so much freedom with how units are handled. We just try to assume what the user meant and "just do it" compared to giving some error message which would indicate the unit mismatch.

Given that this is how it was discussed in the issue, the implementation here is correct. It just makes the implementation much less readable as we assume units and cast to types all over the place.

Reply:
I agree, unfortunately, since we're using a combination of types, I can't see a cleaner approach right away.